### PR TITLE
AP_BattMonitor: use of soc in UAVCAN BatteryInfo

### PR DESCRIPTION
**Pull request by mxiaogit**
Requires the use of soc in UAVCAN BatteryInfo.

### CODE REVIEW

**Comment by magicrub** [comment]
What hardware uses this?

**Review by magicrub — APPROVED** [review]
LGTM but perhaps @WickedShell should sign off on this?

**Comment by mxiaogit** [comment]
> What hardware uses this?
The hardware used is a smart battery, which provides a more accurate SOC.

**Comment by WickedShell** [comment]
@magicrub This will break your battery capacity numbers if using AP_Periph, you will need a follow up PR to populate the the `state_of_health_pct` field (actually looking at it now it looks like the AP_Periph node isn't filling this in correctly to even the unknown state.

**Comment by mxiaogit** [comment]
Do I need to do anything?@WickedShell

**Comment by magicrub** [comment]
This PR looks good to me, I'm happy to merge. Any reason ot hold it back @WickedShell ? I've also created https://github.com/ArduPilot/ardupilot/pull/16016 to populate state_of_charge_pct in AP_Periph

**Review by magicrub — APPROVED** [review]

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp
Comment (tridge) [inline libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:90]: we need to cope with UAVCAN battery monitors that don't know the battery capacity. So either a setting, or cope with zero?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp
Comment (mxiaogit) [reply libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:90]: If the state_of_charge_pct of UAVCAN BatteryInfo is always 0, use AP_BattMonitor_Backend::capacity_remaining_pct().

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp
Comment (magicrub) [inline libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:93]: This Boolean doesn't seem necessary. Can't you do the same check in the capacity accessor function?

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp
Comment (WickedShell) [reply libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:90]: @tridge Do we? The DSDL says that the field is required...

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp
Comment (mxiaogit) [reply libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:93]: When state_of_charge_pct is at [1,100], this boolean is true, indicating that state_of_charge_pct is valid. Until it is true, when state_of_charge_pct is always 0, this boolean is false, indicating that state_of_charge_pct is invalid, use AP_BattMonitor_Backend :: Capacity_remaining_pct(). @magicrub

Comment (mxiaogit) [reply libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:93]: @magicrub @tridge Is there a problem?

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.h
Comment (peterbarker) [inline libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.h:53]: This would be the first use of `mutable` in the ArduPilot codebase.  That would require some discussion - and its not required if you don't assign into _soc_pct (see proposed modification)

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp
Comment (peterbarker) [inline libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:148]:
```suggestion
    if (_use_soc) {
        return _soc;
    }
    return AP_BattMonitor_Backend::capacity_remaining_pct();
```

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp
Comment (peterbarker) [reply libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:93]: @magicrub  I think we do need the persistent boolean - we may give the user readings from 100 down t0 1 - if the device reports 0 we want to pass that on, not switch to the parent class function.

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp
Comment (peterbarker) [reply libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:90]: @tridge standard says one thing, actual devices may say do something else, because life is like that.  Still, perhaps we should assume the field until we do find a counterexample.

**Review by peterbarker — CHANGES_REQUESTED** [review]

**Review by magicrub — APPROVED** [review]

**Review by peterbarker — CHANGES_REQUESTED** [review]
Commit list needs squashing.

I've marked this as DevCall so we can resolve the outstanding questions (and hopefully merge)

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp
Comment (tridge) [reply libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.cpp:93]: by the spec, zero is a valid value

**Review by tridge — CHANGES_REQUESTED** [review]
we need a options bitmask to allow users to say if they want to use the SoC from the node or not. We can't just assume zero is invalid, or that a node will send zero when it doesn't know the capacity

**Comment by mxiaogit** [comment]
If it is always 0, it means the capacity is unknown and AP_BattMonitor_Backend :: capacity_remaining_pct() is used. @tridge

**Comment by tridge** [comment]
> If it is always 0, it means the capacity is unknown and AP_BattMonitor_Backend :: capacity_remaining_pct() is used

always zero does not necessarily mean it is unknown.
Imagine you have 3 batteries in your system, and the uavcan nodes are true smart batteries which really do know the capacity. Then think of the case where the 3rd battery is not connected, so the node is always reporting zero. As there is no current, ArduPilot will report the battery as 100% full, when in fact it isn't connected at all. The user may then takeoff without a battery connected.
We really do need a config option for this. There is just no way to automatically determine the right thing to do.

**Comment by mxiaogit** [comment]
Can you give me some advice on the option bitmask, such as the parameter name.@tridge

**Comment by magicrub** [comment]
@mxiaogit I can do the changes for you real quick and try to push to your branch.

**Comment by mxiaogit** [comment]
@magicrub Thanks for your help.

**Comment by magicrub** [comment]
I've added a branch here: https://github.com/magicrub/ardupilot/tree/cuav_add_capacity_remaining_pct that you can cherry pick commit https://github.com/magicrub/ardupilot/commit/7dce511950d012c31b5425976bd38aeddfa1bbdb over to you branch.

**Comment by mxiaogit** [comment]
@magicrub @WickedShell @tridge @peterbarker Please check the latest submission.

File: libraries/AP_BattMonitor/AP_BattMonitor_Params.h
Comment (tridge) [inline libraries/AP_BattMonitor/AP_BattMonitor_Params.h:20]: this should be an enum class, with value (1U<<1)

**Review by tridge — CHANGES_REQUESTED** [review]
I've put a fix for the issues with this PR here:
https://github.com/tridge/ardupilot/commit/d58d34ce505481e1a09ed0fa1875b4a836f183c2
I can't push it to the PR branch myself as I don't have permission to do that (it is a good idea to allow maintainers to push to PR branches)
I've tested it with this change and it works well.

**Comment by tridge** [comment]
@Hwurzburg once this is merged we will need to document the use of BATT_OPTIONS for controlling how state of charge works for UAVCAN battery monitors

**Review by tridge — APPROVED** [review]

**Review by peterbarker — APPROVED** [review]

**Pull request merged**
